### ISSUE TITLE
variable_names errors

### ISSUE THREAD

**Issue by UWN** [infradig/trealla#234]
<s>

(this is a continuation of the unrelated errors in #231)

Now [#2](http://www.complang.tuwien.ac.at/ulrich/iso-prolog/variable_names#2).

```
ulrich@p0:/opt/gupu/trealla$ git describe
fatal: No annotated tags can describe '6cd6865c6fbc25e0879533f9f2ee3a74871e6676'.
However, there were unannotated tags: try --tags.
ulrich@p0:/opt/gupu/trealla$ make OPT=-std=c99 NOSSL=1
make: Nothing to be done for `all'.
ulrich@p0:/opt/gupu/trealla$ /opt/gupu/trealla/tpl 
Trealla Prolog (c) Infradig 2020-2021, v1.10.20
?- N = 'X', /**/ write_term(T,[quoted(true),variable_names([N=T])]).
uncaught exception: error(instantiation_error,write_term/2)      % unexpected
N = 'X'.
?- N = 'X', /**/ write_term(T,[quoted(true),variable_names([N=T])]), false.
uncaught exception: error(instantiation_error,write_term/2)      % unexpected
false.
?- N = 'X', /**/ catch(write_term(T,[quoted(true),variable_names([N=T])]),error(Err,_),true). 
XN = 'X'.    % where did the error come from before?
?- 
```

There is no error here.
</s>

**Comment by UWN** [comment]
[#30](http://www.complang.tuwien.ac.at/ulrich/iso-prolog/variable_names#30):

```
?- write_term(T,[variable_names(VN_list)]).
uncaught exception: error(domain_error(write_option,variable_names(_7)),write_term/2)   % unexpected, expected: instantiation error
true.
```

**Comment by UWN** [comment]
Please check the [entry for Trealla](http://www.complang.tuwien.ac.at/ulrich/iso-prolog/variable_names). Are the dates etc correct? There are currently 14 problematic cases.

**Comment by infradig** [comment]
The date looks about right.

**Comment by UWN** [comment]
Some 10 left.

Note [#47](http://www.complang.tuwien.ac.at/ulrich/iso-prolog/variable_names#47), which at first glance looks quite reasonable, but here you are not waiting, whereas all other Prologs do wait.

As you have may noticed, the test cases slowly encircle syntax issues. And   [#44](http://www.complang.tuwien.ac.at/ulrich/iso-prolog/variable_names#44) is particularly problematic: Trealla writes a term that will be read back as another term.

Further, I do not understand [#54](http://www.complang.tuwien.ac.at/ulrich/iso-prolog/variable_names#54). Why the instantiation error? There is no instantiation missing at all. Maybe this is more related to the top level than to writer_term/2. Qui sait.

```
ulrich@TU-Wien:/opt/gupu/trealla$ git pull&&make OPT=-std=c99 NOSSL=1&&kill%1;date;git describe;./tpl
Already up to date.
make: Nothing to be done for 'all'.
kill%1: command not found
Di Aug 17 07:55:26 CEST 2021
fatal: No annotated tags can describe '92c9661bed6a75d35a6b7cf916be083e0416ff29'.
However, there were unannotated tags: try --tags.
Trealla Prolog (c) Infradig 2020-2021, v1.10.33
?- S=1+T,N='/*r*/V',/**/ write_term(S,[quoted(true),variable_names([N=T])]). 
uncaught exception: error(instantiation_error,write_term/2)      % unexpected, expected: output 1+/*r*/
S = 1+T,
N = '/*r*/V'.
?- S=1+T,N='/*r*/V',/**/ catch(write_term(S,[quoted(true),variable_names([N=T])]),error(Err,_),true).
1+/*r*/VS = 1+T,                                                 % expected
N = '/*r*/V'.
```

**Comment by infradig** [comment]
With #47, GNU Prolog1.5.0 now gives a domain_error. SWI & Scryer don't as
per your document. Logtalk seems to expect a domain_error.

**Comment by pmoura** [comment]
Test #47 is:

```text
?- VN_list = 42, read_term(T,[variable_names(VN_list)]).
```

If the Prolog system `read_term/3` implementation checks the validity of the options **before** attempting to read a term, a domain error is expected. If the `read_term/3` implementation first attempts to read a term and only **after** that processes the options, then waiting (for a term to be input) is expected. In the former case, the behavior is consistent with other standard predicates such as `open/4` and `write_term/2-3` that also take and check options. In fact, the ISO Prolog Core specification for the `read_term/3` predicate lists a domain error when "An Element `E` of the `Options` list is neither a variable nor a valid read-option".

**Comment by UWN** [comment]
The central question here is: what exactly is a valid read option? For `variable_names/1` it is just the functor alone, no matter the argument. Same with `variables/1` and `singletons/1`. For all those options in 7.10.3, highly operationalizing language is used. ("<b>After inputting</b> a term `VN_list` <b>shall be unified with</b>..).  And never is it stated that these lists etc are part of the option. They are just unified. Whenever explicit unification is used, something operational is happening. Think of 8.4.3.1, 8.5.4.1, 8.10.1.1, 8.12.1.1 ... Here the operational part is the reading itself. Will it take place or not? And the view is: Just read first, unify later.

In the table, there are 2 out of 11 systems that produce this domain error. And those 2 are influenced by the same person that did [not understand the difference between instantiation and domain error](https://github.com/infradig/trealla/issues/231#issuecomment-894804254) which are all defined in 7.12.2 since 1995 nor did read Cor.3:2017. Sorry, I am a bit tired.

**Comment by infradig** [comment]
#44 fixed now.

**Comment by pmoura** [comment]
> The central question here is: what exactly is a valid read option? For variable_names/1 it is just the functor alone, no matter the argument. Same with variables/1 and singletons/1. For all those options in 7.10.3, highly operationalizing language is used. ("After inputting a term VN_list shall be unified with..). And never is it stated that these lists etc are part of the option. They are just unified. Whenever explicit unification is used, something operational is happening. Think of 8.4.3.1, 8.5.4.1, 8.10.1.1, 8.12.1.1 ... Here the operational part is the reading itself. Will it take place or not? And the view is: Just read first, unify later.

Nothing in the ISO Prolog Core standard 8.14.1.1 section, specifically the procedural steps, requires that the check for the validity of the options be performed either before or after a term is read. Doing it before the actual read prevents an error (a non valid option value) from being masked by a failure (when attempting to unify the option value).

> In the table, there are 2 out of 11 systems that produce this domain error. And those 2 are influenced by the same person that did [not understand the difference between instantiation and domain error](https://github.com/infradig/trealla/issues/231#issuecomment-894804254) which are all defined in 7.12.2 since 1995 nor did read (or understand) Cor.3:2017. Sorry, I am a bit tired.

I perfectly understand the difference between an instantiation error and a domain error. That's a quite offensive remark. Maybe actually read what I wrote in the link you provide?

In the original 1995 standard, there was no instantiation error for those cases, only a domain error. Corrigendum 3 changed that in 2017 but Prolog implementations take time to adapt. Hence the interim solution of accepting the two errors. Standards evolve. Test suites for them evolve accordingly. Change doesn't happen instantaneously. Btw, you may want to eventually update your test [table](http://www.complang.tuwien.ac.at/ulrich/iso-prolog/variable_names)? Most of the systems listed there have more recent versions.

As far as my influence goes, Prolog standardization greatly benefited from my lifetime work as it have been publicly acknowledge multiple times, including in WG17 meetings and by multiple Prolog implementers. Maybe the words you're looking for are "Thank you"? You're most welcome.

**Comment by infradig** [comment]
Be that as it may, I don't see how statically type-checking and reporting
an obvious error can ever be wrong.

**Comment by infradig** [comment]
I meant I would like to see such errors reported at compile time, via a
warning.

**Comment by UWN** [comment]
> #44 fixed now.

That fix went too far for two reasons.

1mo, Trealla uses now functional notation which is reserved to the case when there is no operator declaration present or `ignore_ops(true)`. In other words, if there is an operator declaration it must be used (and if there a two, one of them must be used). See a [pasted together version](http://www.complang.tuwien.ac.at/ulrich/iso-prolog/draft-7.10.5) (with all corrections). 

((Because this thread here is full of incorrect claims, note that I quote here just drafts that were used to see the whole picture but which are not the standard. Formally, you need to get everything from ISO (or [cheaper](http://www.complang.tuwien.ac.at/ulrich/iso-prolog/).)))

2do, all prefix operators are currently in functinal notation, or at least to `+a^b` as well. But this is just a special case for minus alone, because minus serves also as a sign for numbers and thus needs some extra handling. See [7.10.5](http://www.complang.tuwien.ac.at/ulrich/iso-prolog/WDCor3#7.10.5_h_2).


(Github makes out of a literal #44 a link to issue #44. You need to quote it in markdown like `[#44]()` which look now like [#44]() - or put the actual link in it)

**Comment by infradig** [comment]
Are talking about this?

:~/trealla$ tpl
?- X = -1,write_term(-X^2,[]), nl.
-(-1^2)

:~/trealla$ scryer
?- X = -1,write_term(-X^2,[]), nl.
- (-1^2)

:~/trealla$ swipl
?- X = -1,write_term(-X^2,[]), nl.
- -1^2

:~/trealla$ gprolog
GNU Prolog 1.5.0 (64 bits)
| ?- X = -1,write_term(-X^2,[]), nl.
- -1^2

**Comment by UWN** [comment]
Yes, there used to be a link in your entry, but there are still plenty of them in [#44](http://www.complang.tuwien.ac.at/ulrich/iso-prolog/variable_names#44) which link to [#183](http://www.complang.tuwien.ac.at/ulrich/iso-prolog/conformity_testing#183).

**Comment by UWN** [comment]
And, the case you quote is a case where there is no ambiguity. But to keep this fix simple, still round brackets are required. Otherwise you have to add some extra contextual information.

**Comment by infradig** [comment]
I'll look at some more when I can.

**Comment by UWN** [comment]
Minus and plus as prefix is used in various cases. Like `-p, +q` mostly denoting some formulas.

**Comment by UWN** [comment]
> Be that as it may, I don't see how statically type-checking and reporting an obvious error can ever be wrong.

There is no static type checking going on. That is the problem. It happens dynamically.

I am unaware of systems that ever produced an error for `read_term(T, [variables(non_list)])` and the like. Also GNU up to 1.4.3 waits for input.

**Comment by infradig** [comment]
The code example *could* be statically typed checked though, and an error
reported.

**Comment by UWN** [comment]
> The code example *could* be statically typed checked though, and an error reported.

You cannot produce your own error during "preparation for execution". You can produce a warning that this built-in will always fail. Even `p :- read_term(_, quelconque).` has to be prepared for execution. How else will you be able to get reproducible results? It would otherwise depend on the intelligence of your analysis. And that will change from release-to-release and from system-to-system.

And we are talking here about a read built-in. Reading is the most sensible side-effecting issue. It is often harder to test than other issues in Prolog. A lot of experience went into the standard, but there are no deep principles to claim for.

Or take it just from another side: Is non-conformance and incompatibility with conforming systems really a price worth paying only to chart out your own idiosyncratic, picky and choosy behavior? And as for GNU Prolog, it did fail after reading the term for `read_term(T, [variables(quidquid)])` up to and including 1.4.3.

I came here because of the very fresh and new approach to `chars`. There is so much more and much more interesting in that direction.  And what about constraints and atvs? For the moment there are only two systems (SICStus and Scryer) which have Holzbaur's interface fully implemented. There is definitely a lot here to think about and to improve.

**Comment by infradig** [comment]
Would love to get back to the good stuff!

**Comment by pmoura** [comment]
> Or take it just from another side: Is non-conformance and incompatibility with conforming systems really a price worth paying only to chart out your own idiosyncratic, picky and choosy behavior?

It's worth reading the ISO Prolog Core standard carefully given your "*this thread here is full of incorrect claims*" comment:

> ((Because this thread here is full of incorrect claims, note that I quote here just drafts that were used to see the whole picture but which are not the standard. Formally, you need to get everything from ISO (or cheaper.)))

Let's start with the specification of the `read_term/3` predicate, section 8.14.1.1 (emphasis mine):

> l) It `T` unifies with `Term`, then **instantiate** the arguments of the read-options (7.10.3) `Options`, and **the goal succeeds**.

Thus, what you claim are "conforming systems" do in a query such as `read_term(T, [variables(non_list)])`? They cannot instantiate the argument of the read-option (as it's already instantiated and the atom `non_list` doesn't unify with any list) and the query fails. **That's non-conforming behavior**.

The 1995 standard in fact stated in section 7.10.3 (emphasis mine):

> `variables(Vars)` - After inputting a term, `Vars` **shall be** a list of the variables in the term input, in left-to-right traversal order.

Cor 3:2017 changes the above sentence to:

> variables(Vars) — After inputting a term, Vars shall be unified with a list of the variables in the term input, in left-to-right traversal order.

But given that the only change to section In 8.14.1.1 in Cor 3:2017 is:

> In 8.14.1.1 k, replace reference 6.4 by 6.2.2.

clause l) is still in effect and only non-contradicting reading of the resulting specification is that a read-option argument that is not a partial list is a domain error as per 8.14.1.3 e) that must be checked before trying to read a term.

> e) An element `E` of the `Options` list is neither a variable nor a valid read-option
> -- `domain_error(read_option, E)`

As a final and general remark, an attempted (and botched) change to the standard that results in masking a programming error with a failure is not progress, it's a bad idea. Also, is no surprise that *traditionally* most Prolog systems didn't check the validity of the read options prior to attempting to read a term as error checking and reporting in general as partial and inconsistent for a long time. Thankfully, more or more systems are becoming less permissive and more strict. Accordingly, that's the direction that is expected for standards, not going back to past, unreliable behavior.

**Comment by UWN** [comment]
> The 1995 standard in fact stated in section 7.10.3 (emphasis mine):
> 
> > `variables(Vars)` - After inputting a term, `Vars` **shall be** a list of the variables in the term input, in left-to-right traversal order.
> 
> Cor 3:2017 changes the above sentence to:
> 
> > variables(Vars) — After inputting a term, Vars shall be unified with a list of the variables in the term input, in left-to-right traversal order.

This is the description of items in the read-options list. Just the next two items immediately following this sentence all read:

> After inputting a term, `VN_list` **shall be unified** ...

From this alone it should be obvious that the correction in the first sentence is a purely editorial one. Not changing any meaning. Or,  should there be a difference between these options? But there is more:

What should be the result to `read_term(T, [variables([])]).` and as text being input just `A.`?  That cannot be a domain error. And it cannot be instantiated in 8.14.1.1 l. But it seems to be that a clean silent failure is appropriate here. My (current) conclusion: there needs to be a fix in 8.14.1.1 alone for that case already. And that means that instead of *instantiates* it needs to add an extra *unifies* into the if-condition.

> > e) An element `E` of the `Options` list is neither a variable nor a valid read-option
> > -- `domain_error(read_option, E)`
> 
> Of course, an `uninstantiation_error/1` would be more appropriated here.

If this would be a new built-in, then this might be a possibility. The need for uninstantiation errors was [first formulated in 2008](http://www.complang.tuwien.ac.at/ulrich/iso-prolog/error_k) and up to now there has not been any suggestion for this particular case. Still, this is a good item to discuss.

**Comment by UWN** [comment]
Quintus Prolog had a range error, which was similar to the `uninstantiation_error` and that  was produced if that unification failed. Search `range_error` [here](http://www.complang.tuwien.ac.at/ulrich/iso-prolog/error_k).

**Comment by pmoura** [comment]
> > The 1995 standard in fact stated in section 7.10.3 (emphasis mine):
> > > `variables(Vars)` - After inputting a term, `Vars` **shall be** a list of the variables in the term input, in left-to-right traversal order.
> > 
> > 
> > Cor 3:2017 changes the above sentence to:
> > > variables(Vars) — After inputting a term, Vars shall be unified with a list of the variables in the term input, in left-to-right traversal order.
> 
> This is the description of items in the read-options list. Just the next two items immediately following this sentence all read:
> 
> > After inputting a term, `VN_list` **shall be unified** ...
> 
> From this alone it should be obvious that the correction in the first sentence is a purely editorial one.

Fully agree. "shall be a list" is a bit odd phrasing given that the argument is necessarily unified (with a list of the variables in the term input). That specific change is an improvement to the standard. That said, the exact behavior of the `read_term/3` doesn't change because of that correction. That is dictated by section 8.14.1.1 and clause l) still implies read-options to be checked prior to attempting to read a term and an error when no valid (i.e. when the argument off the read-option is not unbound).

> Not changing any meaning. Or, should there be a difference between these options? But there is more:
> 
> What should be the result to `read_term(T, [variables([])]).` and as text being input just `A.`? That cannot be a domain error. And it cannot be instantiated in 8.14.1.1 l.

Agree that it should not be a domain error. It should be instead a uninstantiation error. Clause l) doesn't need to be changed but could be phrased better without changing its meaning/semantics.

> But it seems to be that a clean silent failure is appropriate here.

Clause l) disagrees.

> My (current) conclusion: there needs to be a fix in 8.14.1.1 alone for that case already. And that means that instead of _instantiates_ it needs to add an extra _unifies_ into the if-condition.
> 
> > > e) An element `E` of the `Options` list is neither a variable nor a valid read-option
> > > -- `domain_error(read_option, E)`
> > 
> > 
> > Of course, an `uninstantiation_error/1` would be more appropriated here.
> 
> If this would be a new built-in, then this might be a possibility. The need for uninstantiation errors was [first formulated in 2008](http://www.complang.tuwien.ac.at/ulrich/iso-prolog/error_k) and up to now there has not been any suggestion for this particular case. Still, this is a good item to discuss.

Introducing a `uninstantiation_error/1` in 8.14.1.3 after clause e) is the best course of action here. It's an easy change to developers and the only existing code that would "break" should be buggy code where a undiagnosed error have been masked until now by a failure.
